File: Python/Programmars/Level_3/단어 변환.py
def isOneDiff(w1, w2):
    count = 0
    for a, b in zip(w1, w2):
        if a != b:
            count += 1
    if count == 1:
        return True
    return False


def DFS(level, current):
    global answer
    if current == end:
        if level < answer:
            answer = level
    else:
        for i in range(len(w)):
            if check[i] == 0:
                if isOneDiff(w[i], current):
                    check[i] = 1
                    DFS(level+1, w[i])
                    DFS(level, current)
                    check[i] = 0


def solution(begin, target, words):
    if not target in words:
        return 0
    global answer, end, w, check
    answer = len(words)
    end = target
    w = words
    check = [0]*len(words)
    DFS(0, begin)
    return answer


print(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]))  # 4
print(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log"]))  # 0

File: JavaScript/Programmars/고득점-kit/깊이너비 우선 탐색/단어 변환.js
function solution(begin, target, words) {
  let answer = Number.MAX_SAFE_INTEGER;
  const wordsLen = words.length;
  const visited = Array(wordsLen).fill(false);

  // 한 자씩 와일드카드로 바꾼 패턴 생성
  const makePatterns = (word) => {
    return [...word].map((_, i) => {
      return word.slice(0, i) + '.' + word.slice(i + 1);
    });
  };

  function dfs(current, count) {
    if (current === target) {
      answer = Math.min(answer, count);
      return;
    }

    const patterns = makePatterns(current).map((p) => new RegExp(`^${p}$`));

    for (let i = 0; i < wordsLen; i++) {
      if (visited[i]) continue;

      for (let j = 0; j < patterns.length; j++) {
        if (patterns[j].test(words[i])) {
          visited[i] = true;
          dfs(words[i], count + 1);
          visited[i] = false;
        }
      }
    }
  }

  dfs(begin, 0);

  return answer === Number.MAX_SAFE_INTEGER ? 0 : answer;
}

console.log(solution('hit', 'cog', ['hot', 'dot', 'dog', 'lot', 'log', 'cog'])); // 4
console.log(solution('hit', 'cog', ['hot', 'dot', 'dog', 'lot', 'log'])); // 0
